File: Core kernel razonamiento repo para Yaiwes/GitHub-Copilot-SDK/GitHub-Copilot-SDK/nodejs/test/e2e/byok_bearer_token_provider.e2e.test.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import { beforeEach, describe, expect, it } from "vitest";
import { approveAll, CopilotRequestHandler } from "../../src/index.js";
import type {
    CopilotRequestContext,
    BearerTokenProvider,
    NamedProviderConfig,
    ProviderModelConfig,
} from "../../src/index.js";
import { createSdkTestContext } from "./harness/sdkTestContext.js";

/**
 * A captured outbound HTTP request the runtime aimed at a fake BYOK provider
 * endpoint: just the host and the `Authorization` header, which is all these
 * tests need to assert on.
 */
interface CapturedRequest {
    host: string;
    authorization?: string;
}

// Fake BYOK provider base URLs. These hosts are never actually dialed: the
// client-global request interceptor fully answers any request aimed at a
// `.invalid` host, so they only need to be syntactically valid, non-resolving
// URLs. Distinct hosts let the per-provider test assert routing by host.
const PRIMARY_HOST = "byok-endpoint.invalid";
const PRIMARY_BASE_URL = `https://${PRIMARY_HOST}/v1`;
const RED_HOST = "byok-red.invalid";
const RED_BASE_URL = `https://${RED_HOST}/v1`;
const BLUE_HOST = "byok-blue.invalid";
const BLUE_BASE_URL = `https://${BLUE_HOST}/v1`;

/**
 * Client-global HTTP request interceptor (from the SDK's `CopilotRequestHandler`
 * surface) used in place of a real HTTP listener.
 *
 * The runtime invokes {@link sendRequest} for every model-layer HTTP request it
 * would otherwise issue. We capture the ones aimed at a fake BYOK host —
 * recording the `Authorization` header the runtime applied after calling the
 * provider's `bearerTokenProvider` callback over the session-scoped
 * `providerToken.getToken` RPC — and answer them with a synthetic `404` (a
 * non-retryable status, so each outbound model request yields exactly one
 * capture). Every other request (CAPI bootstrap: model catalog, policy, …) is
 * passed straight through to the real network via `super.sendRequest`.
 *
 * Because the handler is client-global (one per CLI process), it is installed
 * once for the whole fixture and {@link reset} between tests.
 */
class CapturingRequestHandler extends CopilotRequestHandler {
    public readonly captures: CapturedRequest[] = [];

    protected override async sendRequest(
        request: Request,
        ctx: CopilotRequestContext
    ): Promise<Response> {
        const url = new URL(request.url);
        if (url.hostname.endsWith(".invalid")) {
            this.captures.push({
                host: url.host,
                authorization: request.headers.get("authorization") ?? undefined,
            });
            return new Response(JSON.stringify({ error: { message: "fake byok endpoint" } }), {
                status: 404,
                headers: { "content-type": "application/json" },
            });
        }
        return super.sendRequest(request, ctx);
    }

    reset(): void {
        this.captures.length = 0;
    }

    /** The `Authorization` headers captured across BYOK requests, in arrival order. */
    authHeaders(): string[] {
        return this.captures
            .map((c) => c.authorization)
            .filter((v): v is string => typeof v === "string");
    }

    /** The `Authorization` header captured for requests aimed at `host`, if any. */
    authHeaderForHost(host: string): string | undefined {
        return this.captures.find((c) => c.host === host)?.authorization;
    }
}

/**
 * End-to-end coverage for the experimental BYOK bearer-token-provider surface
 * (`bearerTokenProvider` on a provider config). The callback stays entirely on the
 * SDK/client side: the SDK strips it from the wire config, sets the
 * `hasBearerTokenProvider` flag, and the runtime calls back over the session-scoped
 * `providerToken.getToken` RPC before each outbound model request, applying the
 * returned token as the `Authorization` header.
 *
 * Rather than standing up a real HTTP listener, these tests install a
 * client-global {@link CapturingRequestHandler} that intercepts the runtime's
 * outbound model request in-process, captures the `Authorization` header, and
 * returns a synthetic response. They validate, against a real runtime:
 *  1. the callback's token reaches the model request as `Authorization: Bearer <token>`;
 *  2. the runtime re-acquires a token per request (no runtime-side caching);
 *  3. per-provider dispatch routes each provider's turn to its own callback,
 *     and the resulting token reaches that provider's endpoint.
 */
describe("BYOK bearer-token provider", async () => {
    const handler = new CapturingRequestHandler();
    const { copilotClient: client } = await createSdkTestContext({
        copilotClientOptions: { requestHandler: handler },
    });

    beforeEach(() => {
        handler.reset();
    });

    /** Drive one BYOK turn; the synthetic 404 errors the turn, which is expected. */
    async function runTurn(
        providers: NamedProviderConfig[],
        models: ProviderModelConfig[],
        selectionId: string,
        prompt: string
    ): Promise<void> {
        const session = await client.createSession({
            onPermissionRequest: approveAll,
            model: selectionId,
            providers,
            models,
        });
        try {
            // The interceptor always 404s, so the turn errors after the runtime
            // has already sent the (token-bearing) request — which is all we
            // assert on. Swallow the resulting error.
            await session.sendAndWait({ prompt }).catch(() => undefined);
        } finally {
            try {
                await session.disconnect();
            } catch {
                // ignore disconnect errors for the fake BYOK endpoint
            }
        }
    }

    it("applies the callback's token as the Authorization header", async () => {
        const SENTINEL = "sentinel-bearer-token-abc123";
        let calls = 0;
        const getBearerToken: BearerTokenProvider = async () => {
            calls += 1;
            return SENTINEL;
        };

        const providers: NamedProviderConfig[] = [
            {
                name: "mi",
                type: "openai",
                wireApi: "completions",
                baseUrl: PRIMARY_BASE_URL,
                bearerTokenProvider: getBearerToken,
            },
        ];
        const models: ProviderModelConfig[] = [
            { id: "default", provider: "mi", wireModel: "byok-gpt-4o" },
        ];

        await runTurn(providers, models, "mi/default", "What is 5+5?");

        // The runtime acquired a token via the callback and applied it verbatim as
        // the bearer credential on the outbound model request.
        expect(handler.authHeaders()).toContain(`Bearer ${SENTINEL}`);
        expect(calls).toBeGreaterThanOrEqual(1);
    });

    it("re-acquires a fresh token for each request (no runtime caching)", async () => {
        let calls = 0;
        const getBearerToken: BearerTokenProvider = async () => {
            calls += 1;
            // A distinct token per acquisition proves the runtime re-invokes the
            // callback per request rather than caching a previous token.
            return `rotating-token-${calls}`;
        };

        const providers: NamedProviderConfig[] = [
            {
                name: "mi",
                type: "openai",
                wireApi: "completions",
                baseUrl: PRIMARY_BASE_URL,
                bearerTokenProvider: getBearerToken,
            },
        ];
        const models: ProviderModelConfig[] = [
            { id: "default", provider: "mi", wireModel: "byok-gpt-4o" },
        ];

        await runTurn(providers, models, "mi/default", "What is 1+1?");
        await runTurn(providers, models, "mi/default", "What is 2+2?");

        // Each outbound request carries a freshly-acquired, distinct token.
        const auths = handler.authHeaders();
        expect(auths.length).toBeGreaterThanOrEqual(2);
        expect(auths[0]).toMatch(/^Bearer rotating-token-\d+$/);
        expect(auths[1]).toMatch(/^Bearer rotating-token-\d+$/);
        expect(auths[0]).not.toBe(auths[1]);
        expect(calls).toBeGreaterThanOrEqual(2);
    });

    it("dispatches token acquisition per provider", async () => {
        const tokenByProvider: Record<string, string> = {
            red: "token-for-red",
            blue: "token-for-blue",
        };
        const acquiredFor: string[] = [];
        const makeCallback =
            (providerName: string): BearerTokenProvider =>
            async (args) => {
                // The runtime forwards the requesting provider's name so the client
                // can dispatch to the right credential.
                expect(args.providerName).toBe(providerName);
                // The runtime also forwards the owning session id so a
                // client-level shared callback can resolve the session.
                expect(typeof args.sessionId).toBe("string");
                expect(args.sessionId.length).toBeGreaterThan(0);
                acquiredFor.push(providerName);
                return tokenByProvider[providerName];
            };

        const providers: NamedProviderConfig[] = [
            {
                name: "red",
                type: "openai",
                wireApi: "completions",
                baseUrl: RED_BASE_URL,
                bearerTokenProvider: makeCallback("red"),
            },
            {
                name: "blue",
                type: "openai",
                wireApi: "completions",
                baseUrl: BLUE_BASE_URL,
                bearerTokenProvider: makeCallback("blue"),
            },
        ];
        const models: ProviderModelConfig[] = [
            { id: "default", provider: "red", wireModel: "byok-gpt-4o" },
            { id: "default", provider: "blue", wireModel: "byok-gpt-4o" },
        ];

        await runTurn(providers, models, "red/default", "What is 3+3?");
        await runTurn(providers, models, "blue/default", "What is 4+4?");

        // Each provider's turn was authenticated with its own token AND that token
        // was delivered to that provider's endpoint, proving per-provider dispatch
        // (not a single session-global credential).
        expect(handler.authHeaderForHost(RED_HOST)).toBe(`Bearer ${tokenByProvider.red}`);
        expect(handler.authHeaderForHost(BLUE_HOST)).toBe(`Bearer ${tokenByProvider.blue}`);
        expect(acquiredFor).toContain("red");
        expect(acquiredFor).toContain("blue");
    });
});

File: ➡️📂Core kernel Yaiwes/GitHub-Copilot-SDK/GitHub-Copilot-SDK/java/sdk/src/test/java/com/github/copilot/ByokBearerTokenProviderE2ETest.java
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

package com.github.copilot;

import static com.github.copilot.CopilotRequestTestSupport.buildNonInferenceResponse;
import static com.github.copilot.CopilotRequestTestSupport.newLlmClient;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.ByteArrayInputStream;
import java.io.InputStream;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpHeaders;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

import javax.net.ssl.SSLSession;

import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import com.github.copilot.rpc.BearerTokenProvider;
import com.github.copilot.rpc.MessageOptions;
import com.github.copilot.rpc.NamedProviderConfig;
import com.github.copilot.rpc.PermissionHandler;
import com.github.copilot.rpc.ProviderModelConfig;
import com.github.copilot.rpc.SessionConfig;

/**
 * End-to-end coverage for the experimental BYOK bearer-token-provider surface
 * ({@code BearerTokenProvider} on a provider config). The callback stays
 * entirely on the SDK/client side: the SDK keeps it off the wire, sends only
 * the {@code hasBearerTokenProvider} flag, and the runtime calls back over the
 * session-scoped {@code providerToken.getToken} RPC before each outbound model
 * request.
 */
public class ByokBearerTokenProviderE2ETest {

    private static final String PRIMARY_HOST = "byok-endpoint.invalid";
    private static final String PRIMARY_BASE_URL = "https://" + PRIMARY_HOST + "/v1";
    private static final String RED_HOST = "byok-red.invalid";
    private static final String RED_BASE_URL = "https://" + RED_HOST + "/v1";
    private static final String BLUE_HOST = "byok-blue.invalid";
    private static final String BLUE_BASE_URL = "https://" + BLUE_HOST + "/v1";

    private static E2ETestContext ctx;
    private CapturingRequestHandler handler;

    @BeforeAll
    static void setup() throws Exception {
        ctx = E2ETestContext.create();
    }

    @AfterAll
    static void teardown() throws Exception {
        if (ctx != null) {
            ctx.close();
        }
    }

    @BeforeEach
    void resetHandler() {
        handler = new CapturingRequestHandler();
    }

    @Test
    void appliesCallbackTokenAsAuthorizationHeader() throws Exception {
        String sentinel = "sentinel-bearer-token-abc123";
        AtomicInteger calls = new AtomicInteger();
        BearerTokenProvider tokenProvider = args -> {
            calls.incrementAndGet();
            return CompletableFuture.completedFuture(sentinel);
        };

        List<NamedProviderConfig> providers = List.of(new NamedProviderConfig().setName("mi").setType("openai")
                .setWireApi("completions").setBaseUrl(PRIMARY_BASE_URL).setBearerTokenProvider(tokenProvider));
        List<ProviderModelConfig> models = List
                .of(new ProviderModelConfig().setId("default").setProvider("mi").setWireModel("byok-gpt-4o"));

        runTurn(providers, models, "mi/default", "What is 5+5?");

        assertTrue(handler.authHeaders().contains("Bearer " + sentinel),
                "Expected captured Authorization headers to contain the callback token: " + handler.authHeaders());
        assertTrue(calls.get() >= 1, "Expected the callback to be invoked at least once");
    }

    @Test
    void reacquiresFreshTokenForEachRequest() throws Exception {
        AtomicInteger calls = new AtomicInteger();
        BearerTokenProvider tokenProvider = args -> CompletableFuture
                .completedFuture("rotating-token-" + calls.incrementAndGet());

        List<NamedProviderConfig> providers = List.of(new NamedProviderConfig().setName("mi").setType("openai")
                .setWireApi("completions").setBaseUrl(PRIMARY_BASE_URL).setBearerTokenProvider(tokenProvider));
        List<ProviderModelConfig> models = List
                .of(new ProviderModelConfig().setId("default").setProvider("mi").setWireModel("byok-gpt-4o"));

        runTurn(providers, models, "mi/default", "What is 1+1?");
        runTurn(providers, models, "mi/default", "What is 2+2?");

        List<String> auths = handler.authHeaders();
        assertTrue(auths.size() >= 2, "Expected at least two captured Authorization headers, got " + auths);
        assertTrue(auths.get(0).startsWith("Bearer rotating-token-"), "Expected rotating token, got " + auths);
        assertTrue(auths.get(1).startsWith("Bearer rotating-token-"), "Expected rotating token, got " + auths);
        assertNotEquals(auths.get(0), auths.get(1), "Expected distinct tokens per request");
        assertTrue(calls.get() >= 2, "Expected the callback to be invoked at least twice");
    }

    @Test
    void dispatchesTokenAcquisitionPerProvider() throws Exception {
        List<String> acquiredFor = new ArrayList<>();
        BearerTokenProvider redCallback = args -> {
            assertEquals("red", args.getProviderName(), "Expected providerName to be forwarded");
            assertTrue(args.getSessionId() != null && !args.getSessionId().isEmpty(),
                    "Expected a non-empty session id in token args");
            synchronized (acquiredFor) {
                acquiredFor.add("red");
            }
            return CompletableFuture.completedFuture("token-for-red");
        };
        BearerTokenProvider blueCallback = args -> {
            assertEquals("blue", args.getProviderName(), "Expected providerName to be forwarded");
            assertTrue(args.getSessionId() != null && !args.getSessionId().isEmpty(),
                    "Expected a non-empty session id in token args");
            synchronized (acquiredFor) {
                acquiredFor.add("blue");
            }
            return CompletableFuture.completedFuture("token-for-blue");
        };

        List<NamedProviderConfig> providers = List.of(
                new NamedProviderConfig().setName("red").setType("openai").setWireApi("completions")
                        .setBaseUrl(RED_BASE_URL).setBearerTokenProvider(redCallback),
                new NamedProviderConfig().setName("blue").setType("openai").setWireApi("completions")
                        .setBaseUrl(BLUE_BASE_URL).setBearerTokenProvider(blueCallback));
        List<ProviderModelConfig> models = List.of(
                new ProviderModelConfig().setId("default").setProvider("red").setWireModel("byok-gpt-4o"),
                new ProviderModelConfig().setId("default").setProvider("blue").setWireModel("byok-gpt-4o"));

        runTurn(providers, models, "red/default", "What is 3+3?");
        runTurn(providers, models, "blue/default", "What is 4+4?");

        assertEquals("Bearer token-for-red", handler.authHeaderForHost(RED_HOST));
        assertEquals("Bearer token-for-blue", handler.authHeaderForHost(BLUE_HOST));
        synchronized (acquiredFor) {
            assertTrue(acquiredFor.contains("red"), "Expected red provider to acquire a token");
            assertTrue(acquiredFor.contains("blue"), "Expected blue provider to acquire a token");
        }
    }

    private void runTurn(List<NamedProviderConfig> providers, List<ProviderModelConfig> models, String selectionId,
            String prompt) throws Exception {
        try (CopilotClient client = newLlmClient(ctx, handler)) {
            CopilotSession session = client
                    .createSession(new SessionConfig().setOnPermissionRequest(PermissionHandler.APPROVE_ALL)
                            .setModel(selectionId).setProviders(providers).setModels(models))
                    .get(60, TimeUnit.SECONDS);
            try {
                session.sendAndWait(new MessageOptions().setPrompt(prompt)).get(60, TimeUnit.SECONDS);
            } catch (Exception ignored) {
                // The fake BYOK endpoint returns 404 after capturing the token-bearing request.
            } finally {
                try {
                    session.close();
                } catch (Exception ignored) {
                    // Ignore disconnect errors for the fake BYOK endpoint.
                }
            }
        }
    }

    private static final class CapturingRequestHandler extends CopilotRequestHandler {

        private final ConcurrentLinkedQueue<CapturedRequest> captures = new ConcurrentLinkedQueue<>();

        @Override
        protected HttpResponse<InputStream> sendRequest(HttpRequest request, CopilotRequestContext rctx)
                throws Exception {
            String host = request.uri().getHost();
            if (host != null && host.endsWith(".invalid")) {
                captures.add(new CapturedRequest(request.uri().getHost(),
                        request.headers().firstValue("Authorization").orElse(null)));
                return new StubHttpResponse(404, "{\"error\":{\"message\":\"fake byok endpoint\"}}");
            }
            return buildNonInferenceResponse(request.uri().toString());
        }

        List<String> authHeaders() {
            List<String> auths = new ArrayList<>();
            for (CapturedRequest capture : captures) {
                if (capture.authorization() != null) {
                    auths.add(capture.authorization());
                }
            }
            return auths;
        }

        String authHeaderForHost(String host) {
            for (CapturedRequest capture : captures) {
                if (host.equals(capture.host())) {
                    return capture.authorization();
                }
            }
            return null;
        }
    }

    private static final class StubHttpResponse implements HttpResponse<InputStream> {

        private final int status;
        private final HttpHeaders headers;
        private final byte[] body;

        StubHttpResponse(int status, String body) {
            this.status = status;
            this.body = body.getBytes(StandardCharsets.UTF_8);
            this.headers = HttpHeaders.of(Map.of("content-type", List.of("application/json")), (k, v) -> true);
        }

        @Override
        public int statusCode() {
            return status;
        }

        @Override
        public HttpRequest request() {
            return null;
        }

        @Override
        public Optional<HttpResponse<InputStream>> previousResponse() {
            return Optional.empty();
        }

        @Override
        public HttpHeaders headers() {
            return headers;
        }

        @Override
        public InputStream body() {
            return new ByteArrayInputStream(body);
        }

        @Override
        public Optional<SSLSession> sslSession() {
            return Optional.empty();
        }

        @Override
        public URI uri() {
            return null;
        }

        @Override
        public HttpClient.Version version() {
            return HttpClient.Version.HTTP_1_1;
        }
    }

    private record CapturedRequest(String host, String authorization) {
    }
}
